src/envs/repeated_nextto_painting.py

```diff
@@ -0,0 +1,182 @@
+        if held_obj is not None:
+            next_state.set(held_obj, "pose_y", y)
+        return next_state
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {
+            self._InBox, self._InShelf, self._IsBoxColor, self._IsShelfColor,
+            self._GripperOpen, self._OnTable, self._HoldingTop,
+            self._HoldingSide, self._Holding, self._IsWet, self._IsDry,
+            self._IsDirty, self._IsClean, self._NextTo, self._NextToBox,
+            self._NextToShelf, self._NextToNothing
+        }
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {
+            self._Pick, self._Wash, self._Dry, self._Paint, self._Place,
+            self._OpenLid, self._MoveToObj, self._MoveToBox, self._MoveToShelf
+        }
+
+    @property
+    def action_space(self) -> Box:
+        # Actions are 8-dimensional vectors:
+        # [x, y, z, grasp, pickplace, water level, heat level, color]
+        # Note that pickplace is 1 for pick, -1 for place, and 0 otherwise,
+        # while grasp, water level, heat level, and color are in [0, 1].
+        # Changed lower bound for z to 0.0 for RepeatedNextToPainting
+        # This is needed to check affinity of the move action
+        lowers = np.array(
+            [self.obj_x - 1e-2, self.env_lb, 0.0, 0.0, -1.0, 0.0, 0.0, 0.0],
+            dtype=np.float32)
+        uppers = np.array([
+            self.obj_x + 1e-2, self.env_ub, self.obj_z + 1e-2, 1.0, 1.0, 1.0,
+            1.0, 1.0
+        ],
+                          dtype=np.float32)
+        return Box(lowers, uppers)
+
+    def _OnTable_holds(self, state: State, objects: Sequence[Object]) -> bool:
+        obj, = objects
+        obj_y = state.get(obj, "pose_y")
+        return self.table_lb < obj_y < self.table_ub and \
+            np.allclose(state.get(obj, "pose_z"), self.table_height\
+                + self.obj_height / 2)
+
+    @staticmethod
+    def _Move_policy(state: State, memory: Dict, objects: Sequence[Object],
+                     params: Array) -> Action:
+        del memory  # unused
+        _, obj = objects
+        next_x = state.get(obj, "pose_x")
+        next_y = params[0]
+        return Action(
+            np.array([next_x, next_y, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+                     dtype=np.float32))
+
+    def _NextTo_holds(self, state: State, objects: Sequence[Object]) -> bool:
+        robot, obj = objects
+        return abs(state.get(robot, "y") -
+                   state.get(obj, "pose_y")) < self.nextto_thresh
+
+    def _NextToNothing_holds(self, state: State,
+                             objects: Sequence[Object]) -> bool:
+        robot, = objects
+        for typed_obj in state:
+            if typed_obj.type in \
+                [self._obj_type, self._box_type, self._shelf_type] and \
+                self._NextTo_holds(state, [robot, typed_obj]) and \
+                typed_obj is not self._get_held_object(state):
```

Comment:
can we check first at the beginning of this function once whether anything is held?